### ISSUE TITLE
data.aws_iam_policy_document problems in 0.7.9

### ISSUE THREAD

**Issue by cread** [hashicorp/terraform#9977]
The following config works perfectly in `0.7.7` but has problems in `0.7.9`

```hcl
variable "bucket_name" {
  default = "test.log"
}

variable "acl" {
  default = "log-delivery-write"
}

resource "aws_s3_bucket" "bucket" {
  bucket = "${var.bucket_name}"
  acl    = "${var.acl}"
}

data "aws_iam_policy_document" "policy" {
  statement {
    sid = "DenyDeletingObjects"
    effect = "Deny"
    actions = [
      "s3:DeleteObject",
      "s3:DeleteObjectVersion"
    ]
    principals {
      type = "AWS"
      identifiers = ["*"]
    }
    resources = [
      "${aws_s3_bucket.bucket.arn}/*"
    ]
  }

  statement {
    sid = "DenyDeletingBucket"
    effect = "Deny"
    actions = [
      "s3:DeleteBucket"
    ]
    principals {
      type = "AWS"
      identifiers = ["*"]
    }
    resources = [
      "${aws_s3_bucket.bucket.arn}"
    ]
  }
}

resource "aws_s3_bucket_policy" "bucket"{
  bucket = "${aws_s3_bucket.bucket.bucket}"
  policy = "${data.aws_iam_policy_document.policy.json}"
}
```

With no `terraform.tfstate` file, the `terraform plan` command gives the following error:

```
1 error(s) occurred:

* plan operation: ReadDataDiff "data.aws_iam_policy_document.policy" had unequal configs (real, then shadow):

&terraform.ResourceConfig{ComputedKeys:[]string{"statement.resources", "statement.resources"}, Raw:map[string]interface {}{"statement":[]map[string]interface {}{map[string]interface {}{"actions":[]interface {}{"s3:DeleteObject", "s3:DeleteObjectVersion"}, "principals":[]map[string]interface {}{map[string]interface {}{"type":"AWS", "identifiers":[]interface {}{"*"}}}, "resources":[]interface {}{"${aws_s3_bucket.bucket.arn}/*"}, "sid":"DenyDeletingObjects", "effect":"Deny"}, map[string]interface {}{"sid":"DenyDeletingBucket", "effect":"Deny", "actions":[]interface {}{"s3:DeleteBucket"}, "principals":[]map[string]interface {}{map[string]interface {}{"type":"AWS", "identifiers":[]interface {}{"*"}}}, "resources":[]interface {}{"${aws_s3_bucket.bucket.arn}"}}}}, Config:map[string]interface {}{"statement":[]map[string]interface {}{map[string]interface {}{"principals":[]map[string]interface {}{map[string]interface {}{"type":"AWS", "identifiers":[]interface {}{"*"}}}, "sid":"DenyDeletingObjects", "effect":"Deny", "actions":[]interface {}{"s3:DeleteObject", "s3:DeleteObjectVersion"}}, map[string]interface {}{"principals":[]map[string]interface {}{map[string]interface {}{"identifiers":[]interface {}{"*"}, "type":"AWS"}}, "sid":"DenyDeletingBucket", "effect":"Deny", "actions":[]interface {}{"s3:DeleteBucket"}}}}, raw:(*config.RawConfig)(0xc4204f5a40)}

&terraform.ResourceConfig{ComputedKeys:[]string{"statement.1.resources", "statement.resources"}, Raw:map[string]interface {}{"statement":[]map[string]interface {}{map[string]interface {}{"actions":[]interface {}{"s3:DeleteObject", "s3:DeleteObjectVersion"}, "principals":[]map[string]interface {}{map[string]interface {}{"type":"AWS", "identifiers":[]interface {}{"*"}}}, "resources":[]interface {}{"${aws_s3_bucket.bucket.arn}/*"}, "sid":"DenyDeletingObjects", "effect":"Deny"}, map[string]interface {}{"sid":"DenyDeletingBucket", "effect":"Deny", "actions":[]interface {}{"s3:DeleteBucket"}, "principals":[]map[string]interface {}{map[string]interface {}{"type":"AWS", "identifiers":[]interface {}{"*"}}}, "resources":[]interface {}{"${aws_s3_bucket.bucket.arn}"}}}}, Config:map[string]interface {}{"statement":[]map[string]interface {}{map[string]interface {}{"sid":"DenyDeletingObjects", "effect":"Deny", "actions":[]interface {}{"s3:DeleteObject", "s3:DeleteObjectVersion"}, "principals":[]map[string]interface {}{map[string]interface {}{"type":"AWS", "identifiers":[]interface {}{"*"}}}}, map[string]interface {}{"effect":"Deny", "actions":[]interface {}{"s3:DeleteBucket"}, "principals":[]map[string]interface {}{map[string]interface {}{"type":"AWS", "identifiers":[]interface {}{"*"}}}, "sid":"DenyDeletingBucket"}}}, raw:(*config.RawConfig)(0xc42014cf00)}
```

**Comment by mitchellh** [comment]
This is fixed and will be part of 0.7.10 tomorrow!

**Comment by cread** [comment]
Sadly this is not the case.

I've just tried again with `v0.7.10` and get the following error:

```
* plan operation: ReadDataDiff "data.aws_iam_policy_document.policy" had unequal configs (real, then shadow):

&terraform.ResourceConfig{ComputedKeys:[]string{"statement.resources", "statement.resources"}, Raw:map[string]interface {}{"statement":[]map[string]interface {}{map[string]interface {}{"resources":[]interface {}{"${aws_s3_bucket.bucket.arn}/*"}, "sid":"DenyDeletingObjects", "effect":"Deny", "actions":[]interface {}{"s3:DeleteObject", "s3:DeleteObjectVersion"}, "principals":[]map[string]interface {}{map[string]interface {}{"type":"AWS", "identifiers":[]interface {}{"*"}}}}, map[string]interface {}{"sid":"DenyDeletingBucket", "effect":"Deny", "actions":[]interface {}{"s3:DeleteBucket"}, "principals":[]map[string]interface {}{map[string]interface {}{"type":"AWS", "identifiers":[]interface {}{"*"}}}, "resources":[]interface {}{"${aws_s3_bucket.bucket.arn}"}}}}, Config:map[string]interface {}{"statement":[]map[string]interface {}{map[string]interface {}{"effect":"Deny", "actions":[]interface {}{"s3:DeleteObject", "s3:DeleteObjectVersion"}, "principals":[]map[string]interface {}{map[string]interface {}{"identifiers":[]interface {}{"*"}, "type":"AWS"}}, "sid":"DenyDeletingObjects"}, map[string]interface {}{"actions":[]interface {}{"s3:DeleteBucket"}, "principals":[]map[string]interface {}{map[string]interface {}{"type":"AWS", "identifiers":[]interface {}{"*"}}}, "sid":"DenyDeletingBucket", "effect":"Deny"}}}, raw:(*config.RawConfig)(0xc420385ce0)}

&terraform.ResourceConfig{ComputedKeys:[]string{"statement.1.resources", "statement.resources"}, Raw:map[string]interface {}{"statement":[]map[string]interface {}{map[string]interface {}{"actions":[]interface {}{"s3:DeleteObject", "s3:DeleteObjectVersion"}, "principals":[]map[string]interface {}{map[string]interface {}{"type":"AWS", "identifiers":[]interface {}{"*"}}}, "resources":[]interface {}{"${aws_s3_bucket.bucket.arn}/*"}, "sid":"DenyDeletingObjects", "effect":"Deny"}, map[string]interface {}{"sid":"DenyDeletingBucket", "effect":"Deny", "actions":[]interface {}{"s3:DeleteBucket"}, "principals":[]map[string]interface {}{map[string]interface {}{"type":"AWS", "identifiers":[]interface {}{"*"}}}, "resources":[]interface {}{"${aws_s3_bucket.bucket.arn}"}}}}, Config:map[string]interface {}{"statement":[]map[string]interface {}{map[string]interface {}{"actions":[]interface {}{"s3:DeleteObject", "s3:DeleteObjectVersion"}, "principals":[]map[string]interface {}{map[string]interface {}{"type":"AWS", "identifiers":[]interface {}{"*"}}}, "sid":"DenyDeletingObjects", "effect":"Deny"}, map[string]interface {}{"sid":"DenyDeletingBucket", "effect":"Deny", "actions":[]interface {}{"s3:DeleteBucket"}, "principals":[]map[string]interface {}{map[string]interface {}{"type":"AWS", "identifiers":[]interface {}{"*"}}}}}}, raw:(*config.RawConfig)(0xc4200f6a80)}
```

**Comment by mitchellh** [comment]
Whoops! Reopening to take a look. Thanks @cread

**Comment by mitchellh** [comment]
This will be fixed by #10118

Closing this, please subscribe to #10075 for more info. That is the issue that we just ended up centralizing on.
